### PR TITLE
feat: [DestinationService] Allow optional retrieval-strategy (tenant) in getAllDestinationProperties

### PR DESCRIPTION
**Pull request by newtork**
## Context
 

#558

We deprecated public `DestinationService` method `public Try<Iterable<Destination>> tryGetAllDestinations( @Nonnull final DestinationOptions options )`. The user wants an alternative to ensure the destination lookup happens for correct tenant.

## Definition of Done

- [x] Functionality scope stated & covered
- [x] Tests cover the scope above
- [x] ~Error handling created / updated & covered by the tests above~
- [x] ~Documentation updated~
- [x] Release notes updated

### CODE REVIEW

File: cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java [hunk 1]
```diff
@@ -264,7 +286,7 @@ private Try<List<DestinationProperties>> getAllDestinationsByRetrievalStrategy(
             DestinationRetrievalStrategyResolver
                 .forAllDestinations(adapter::getProviderTenantId, this::loadAndParseAllDestinations);
 
-        return Try.success(options).map(resolver::prepareSupplierAllDestinations).map(Supplier::get);
+        return Try.of(() -> resolver.prepareSupplierAllDestinations(options).get());
```
Comment (newtork) [inline cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java:289]: (Comment)

Code style, drive-by.

File: cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java [hunk 2]
```diff
@@ -184,21 +184,42 @@ public Try<Iterable<Destination>> tryGetAllDestinations()
      * destination at service instance level takes precedence.
      *
      * @return A list of destination properties.
+     * @see DestinationService#getAllDestinationProperties(DestinationServiceRetrievalStrategy)
      * @since 5.1.0
      */
     @Nonnull
-    public Collection<DestinationProperties> getAllDestinationProperties()
+    public List<DestinationProperties> getAllDestinationProperties()
```
Comment (newtork) [inline cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java:191]: (Comment)

I forgot why we decided for `Collection` instead of `List`.
But I reckon, the latter would be easier to use, right?

File: cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java [hunk 3]
```diff
@@ -184,21 +184,42 @@ public Try<Iterable<Destination>> tryGetAllDestinations()
...
+     * In case there exists a destination with the same name on service instance and on sub-account level, the
+     * destination at service instance level takes precedence.
+     *
+     * @return A list of destination properties.
+     * @param retrievalStrategy
+     *            Strategy for loading destinations in a multi-tenant application.
+     * @since 5.12.0
+     */
+    @Nonnull
+    public List<DestinationProperties> getAllDestinationProperties(
```
Comment (MatKuhr) [inline cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java:212]: This also raises the question how we would handle requests for further parameters, e.g. retrieving only instance or only subaccount destinations, or `$filter`, `$select` etc.

File: cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java (hunk 1 above)
Comment (MatKuhr) [reply cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java:289]: (minor) Not sure this improves readability? I like about the left side that it's a bit more obvious what the return type of `prepareSupplierAllDestinations` is. But no strong preference

File: cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java [hunk 4]
```diff
@@ -467,6 +467,50 @@ void testGettingDestinationProperties()
...
+        doReturn(responseSubaccountDestination)
+            .when(destinationServiceAdapter)
+            .getConfigurationAsJson("/subaccountDestinations", withoutToken(TECHNICAL_USER_PROVIDER));
+
+        final Collection<DestinationProperties> destinationList = loader.getAllDestinationProperties(ALWAYS_PROVIDER);
+        assertThat(destinationList)
+            .extracting(d -> d.get(DestinationProperty.NAME).get())
+            .containsExactly("CC8-HTTP-BASIC", "CC8-HTTP-CERT1", "CC8-HTTP-CERT");
+
+        // verify all results are cached
```
Comment (MatKuhr) [inline cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java:485]: how does this test caching, if the above code is only invoked once?

File: cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java [hunk 5]
```diff
@@ -584,6 +628,22 @@ void testGetAllDestinationsOnlySubscriberStrategyReadsSubscriberDestinations()
...
+        // set current tenant to be the provider tenant
+        context.clearTenant();
+
+        TenantAccessor
+            .executeWithTenant(
+                providerTenant,
+                () -> assertThatThrownBy(() -> loader.getAllDestinationProperties(ONLY_SUBSCRIBER))
+                    .isInstanceOf(DestinationAccessException.class)
+                    .hasMessageContaining(
+                        "The current tenant is the provider tenant, which should not be the case with the option OnlySubscriber. Cannot retrieve destination."));
```
Comment (MatKuhr) [inline cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java:644]: This seems redundant, the following should work:

```suggestion
        context.setTenant(providerTenant);
        assertThatThrownBy(() -> loader.getAllDestinationProperties(ONLY_SUBSCRIBER))
                    .isInstanceOf(DestinationAccessException.class)
                    .hasMessageContaining(
                        "The current tenant is the provider tenant, which should not be the case with the option OnlySubscriber. Cannot retrieve destination.");
```

**Review by MatKuhr — COMMENTED** [review]
essentially LGTM, some minor questions

File: cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java (hunk 1 above)
Comment (newtork) [reply cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java:289]: I will revert. I mostly hated that the line is starting with `Try.success(options).`

File: cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java (hunk 2 above)
Comment (newtork) [reply cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java:191]: I will revert.

File: cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java (hunk 3 above)
Comment (newtork) [reply cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java:212]: While I agree with the uncertainty of future retrieval options for `$filter`, `$select`, etc. - the method is called "getAllDestinationProperties". For a reduced result set we will most likely consider a new method name (and signature) anyway.

File: cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java (hunk 4 above)
Comment (newtork) [reply cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java:485]: It was a lazy copy-paste mistake

File: cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java (hunk 5 above)
Comment (newtork) [reply cloudplatform/connectivity-destination-service/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationServiceTest.java]: Works well!

**Review by MatKuhr — COMMENTED** [review]
Adding a test that asserts the cache respects this tenant setting might be useful? Not sure how obvious this is from the implementation..

Otherwise LGTM

**Review by MatKuhr — APPROVED** [review]
